File: introduction/csharp/src/Smelly.Code.Core/EvercraftGame.cs
using System;

namespace Smelly.Code.Core
{
    public class EvercraftGame
    {
        private int?[] Strength { get; } = {null, null};
        private int?[] Dexterity { get; } = {null, null};
        public Character[] Chars { get; set; }

        public void Start()
        {
            Chars = new[]
            {
                new Character(5, 10),
                new Character(5, 10),
            };
        }

        public void Attack(int roll, Character character)
        {
            var sM = 0;
            var dM = 0;
            var charIndex = Array.IndexOf(Chars, character);

            if (charIndex == 0)
            {
                if (Strength[0] == 1)
                {
                    sM = -5;
                }
                else if (Strength[0] == 2 || Strength[0] == 3)
                {
                    sM = -4;
                }
                else if (Strength[0] == 4 || Strength[0] == 5)
                {
                    sM = -3;
                }
                else if (Strength[0] == 6 || Strength[0] == 7)
                {
                    sM = -2;
                }
                else if (Strength[0] == 8 || Strength[0] == 9)
                {
                    sM = -1;
                }
                else if (Strength[0] == 12 || Strength[0] == 13)
                {
                    sM = 1;
                }
                else if (Strength[0] == 14 || Strength[0] == 15)
                {
                    sM = 2;
                }
                else if (Strength[0] == 16 || Strength[0] == 17)
                {
                    sM = 3;
                }
                else if (Strength[0] == 18 || Strength[0] == 19)
                {
                    sM = 4;
                }
                else if (Strength[0] == 20)
                {
                    sM = 5;
                }

                if (Dexterity[1] == 1)
                {
                    dM = -5;
                }
                else if (Dexterity[1] == 2 || Dexterity[1] == 3)
                {
                    dM = -4;
                }
                else if (Dexterity[1] == 4 || Dexterity[1] == 5)
                {
                    dM = -3;
                }
                else if (Dexterity[1] == 6 || Dexterity[1] == 7)
                {
                    dM = -2;
                }
                else if (Dexterity[1] == 8 || Dexterity[1] == 9)
                {
                    dM = -1;
                }
                else if (Dexterity[1] == 12 || Dexterity[1] == 13)
                {
                    dM = 1;
                }
                else if (Dexterity[1] == 14 || Dexterity[1] == 15)
                {
                    dM = 2;
                }
                else if (Dexterity[1] == 16 || Dexterity[1] == 17)
                {
                    dM = 3;
                }
                else if (Dexterity[1] == 18 || Dexterity[1] == 19)
                {
                    dM = 4;
                }
                else if (Dexterity[1] == 20)
                {
                    dM = 5;
                }

                if (roll + sM >= Chars[1].Armor + dM)
                {
                    Chars[1].HitPoints = Chars[1].HitPoints - 1;
                }

                if (Strength[0].HasValue)
                {
                    if (sM > 0 && roll + sM >= Chars[1].Armor)
                    {
                        Chars[1].HitPoints = Chars[1].HitPoints - sM;
                    }
                    else
                    {
                        Chars[1].HitPoints = Chars[1].HitPoints - 1;
                    }
                }


                if (roll == 20)
                {
                    Chars[1].HitPoints = Chars[1].HitPoints - 1;
                }
            }
            else
            {
                if (Strength[1] == 1)
                {
                    sM = -5;
                }
                else if (Strength[1] == 2 || Strength[1] == 3)
                {
                    sM = -4;
                }
                else if (Strength[1] == 4 || Strength[1] == 5)
                {
                    sM = -3;
                }
                else if (Strength[1] == 6 || Strength[1] == 7)
                {
                    sM = -2;
                }
                else if (Strength[1] == 8 || Strength[1] == 9)
                {
                    sM = -1;
                }
                else if (Strength[1] == 12 || Strength[1] == 13)
                {
                    sM = 1;
                }
                else if (Strength[1] == 14 || Strength[1] == 15)
                {
                    sM = 2;
                }
                else if (Strength[1] == 16 || Strength[1] == 17)
                {
                    sM = 3;
                }
                else if (Strength[1] == 18 || Strength[1] == 19)
                {
                    sM = 4;
                }
                else if (Strength[1] == 20)
                {
                    sM = 5;
                }
                
                if (Dexterity[0] == 1)
                {
                    dM = -5;
                }
                else if (Dexterity[0] == 2 || Dexterity[0] == 3)
                {
                    dM = -4;
                }
                else if (Dexterity[0] == 4 || Dexterity[0] == 5)
                {
                    dM = -3;
                }
                else if (Dexterity[0] == 6 || Dexterity[0] == 7)
                {
                    dM = -2;
                }
                else if (Dexterity[0] == 8 || Dexterity[0] == 9)
                {
                    dM = -1;
                }
                else if (Dexterity[0] == 12 || Dexterity[0] == 13)
                {
                    dM = 1;
                }
                else if (Dexterity[0] == 14 || Dexterity[0] == 15)
                {
                    dM = 2;
                }
                else if (Dexterity[0] == 16 || Dexterity[0] == 17)
                {
                    dM = 3;
                }
                else if (Dexterity[0] == 18 || Dexterity[0] == 19)
                {
                    dM = 4;
                }
                else if (Dexterity[0] == 20)
                {
                    dM = 5;
                }
                
                if (roll + sM >= Chars[0].Armor + dM)
                {
                    Chars[0].HitPoints = Chars[0].HitPoints - 1;
                }
                
                if (Strength[1].HasValue)
                {
                    if (sM > 0 && roll + sM >= Chars[0].Armor)
                    {
                        Chars[0].HitPoints = Chars[0].HitPoints - sM;
                    }
                    else
                    {
                        Chars[0].HitPoints = Chars[0].HitPoints - 1;
                    }
                }
            }
        }

        public bool IsDead(Character character)
        {
            return character.HitPoints == 0;
        }

        public void EquipArmor(ArmorType armorType, Character character)
        {
            switch (armorType)
            {
                case ArmorType.Bronze:
                    character.Armor = character.Armor - 1;
                    break;
                case ArmorType.Steel:
                    character.Armor = character.Armor + 1;
                    break;
            }
        }

        public void ApplyStrength(int strength, Character character)
        {
            var charIndex = Array.IndexOf(Chars, character);
            Strength[charIndex] = strength;
        }

        public void ApplyDexterity(int dexterity, Character character)
        {
            var charIndex = Array.IndexOf(Chars, character);
            Dexterity[charIndex] = dexterity;
        }
    }
}

File: javascript/src/evercraft-game.js
import { Character } from './character';
import { ARMOR_TYPE } from './armor-type';
import fs from 'fs';
import sqlite from 'sqlite';

export class EvercraftGame {
  attacked = [false, false];
  chars = [null, null];

  start = (charOne = 'Jack', charTwo = 'Bob') => {
    this.chars = [
      new Character(charOne, 5, 10),
      new Character(charTwo, 5, 10)
    ];
  };

  attack = (roll, char) => {
    let sM = 0;
    let dM = 0;
    const charIndex = this.chars.indexOf(char);

    if (charIndex !== 0) {
      this.attacked[1] = true;
      if (this.chars[0].str === 1) {
        sM = -5;
      } else if (this.chars[0].str == 2 || this.chars[0].str == 3) {
        sM = -4;
      } else if (this.chars[0].str == 4 || this.chars[0].str == 5) {
        sM = -3;
      } else if (this.chars[0].str == 6 || this.chars[0].str == 7) {
        sM = -2;
      } else if (this.chars[0].str == 8 || this.chars[0].str == 9) {
        sM = -1;
      } else if (this.chars[0].str == 12 || this.chars[0].str == 13) {
        sM = 1;
      } else if (this.chars[0].str == 14 || this.chars[0].str == 15) {
        sM = 2;
      } else if (this.chars[0].str == 16 || this.chars[0].str == 17) {
        sM = 3;
      } else if (this.chars[0].str == 18 || this.chars[0].str == 19) {
        sM = 4;
      } else if (this.chars[0].str == 20) {
        sM = 5;
      }

      if (this.chars[1].dex == 1) {
        dM = -5;
      } else if (this.chars[1].dex == 2 || this.chars[1].dex == 3) {
        dM = -4;
      } else if (this.chars[1].dex == 4 || this.chars[1].dex == 5) {
        dM = -3;
      } else if (this.chars[1].dex == 6 || this.chars[1].dex == 7) {
        dM = -2;
      } else if (this.chars[1].dex == 8 || this.chars[1].dex == 9) {
        dM = -1;
      } else if (this.chars[1].dex == 12 || this.chars[1].dex == 13) {
        dM = 1;
      } else if (this.chars[1].dex == 14 || this.chars[1].dex == 15) {
        dM = 2;
      } else if (this.chars[1].dex == 16 || this.chars[1].dex == 17) {
        dM = 3;
      } else if (this.chars[1].dex == 18 || this.chars[1].dex == 19) {
        dM = 4;
      } else if (this.chars[1].dex == 20) {
        dM = 5;
      }

      if (roll + sM >= this.chars[1].arm + dM) {
        this.chars[1].hitPts = this.chars[1].hitPts - 1;
      }

      if (this.chars[0].str) {
        if (sM > 0 && roll + sM >= this.chars[1].arm) {
          this.chars[1].hitPts = this.chars[1].hitPts - sM;
        } else {
          this.chars[1].hitPts = this.chars[1].hitPts - 1;
        }
      }


      if (roll == 20) {
        this.chars[1].hitPts = this.chars[1].hitPts - 1;
      }
    } else {
      this.attacked[0] = true;
      if (this.chars[1].str == 1) {
        sM = -5;
      } else if (this.chars[1].str == 2 || this.chars[1].str == 3) {
        sM = -4;
      } else if (this.chars[1].str == 4 || this.chars[1].str == 5) {
        sM = -3;
      } else if (this.chars[1].str == 6 || this.chars[1].str == 7) {
        sM = -2;
      } else if (this.chars[1].str == 8 || this.chars[1].str == 9) {
        sM = -1;
      } else if (this.chars[1].str == 12 || this.chars[1].str == 13) {
        sM = 1;
      } else if (this.chars[1].str == 14 || this.chars[1].str == 15) {
        sM = 2;
      } else if (this.chars[1].str == 16 || this.chars[1].str == 17) {
        sM = 3;
      } else if (this.chars[1].str == 18 || this.chars[1].str == 19) {
        sM = 4;
      } else if (this.chars[1].str == 20) {
        sM = 5;
      }

      if (this.chars[0].dex == 1) {
        dM = -5;
      } else if (this.chars[0].dex == 2 || this.chars[0].dex == 3) {
        dM = -4;
      } else if (this.chars[0].dex == 4 || this.chars[0].dex == 5) {
        dM = -3;
      } else if (this.chars[0].dex == 6 || this.chars[0].dex == 7) {
        dM = -2;
      } else if (this.chars[0].dex == 8 || this.chars[0].dex == 9) {
        dM = -1;
      } else if (this.chars[0].dex == 12 || this.chars[0].dex == 13) {
        dM = 1;
      } else if (this.chars[0].dex == 14 || this.chars[0].dex == 15) {
        dM = 2;
      } else if (this.chars[0].dex == 16 || this.chars[0].dex == 17) {
        dM = 3;
      } else if (this.chars[0].dex == 18 || this.chars[0].dex == 19) {
        dM = 4;
      } else if (this.chars[0].dex == 20) {
        dM = 5;
      }

      if (roll + sM >= this.chars[0].arm + dM) {
        this.chars[0].hitPts = this.chars[0].hitPts - 1;
      }

      if (this.chars[1].str) {
        if (sM > 0 && roll + sM >= this.chars[0].arm) {
          this.chars[0].hitPts = this.chars[0].hitPts - sM;
        } else {
          this.chars[0].hitPts = this.chars[0].hitPts - 1;
        }
      }
    }
  };

  isDead = (character) => {
    const charIndex = this.chars.indexOf(character);
    const hitPoints = character.hitPts;
    let hM = 0;
    if (this.chars[charIndex].cst) {
      if (this.chars[charIndex].cst == 1) {
        hM = -5;
      } else if (this.chars[charIndex].cst == 2 || this.chars[charIndex].cst == 3) {
        hM = -4;
      } else if (this.chars[charIndex].cst == 4 || this.chars[charIndex].cst == 5) {
        hM = -3;
      } else if (this.chars[charIndex].cst == 6 || this.chars[charIndex].cst == 7) {
        hM = -2;
      } else if (this.chars[charIndex].cst == 8 || this.chars[charIndex].cst == 9) {
        hM = -1;
      } else if (this.chars[charIndex].cst == 12 || this.chars[charIndex].cst == 13) {
        hM = 1;
      } else if (this.chars[charIndex].cst == 14 || this.chars[charIndex].cst == 15) {
        hM = 2;
      } else if (this.chars[charIndex].cst == 16 || this.chars[charIndex].cst == 17) {
        hM = 3;
      } else if (this.chars[charIndex].cst == 18 || this.chars[charIndex].cst == 19) {
        hM = 4;
      } else if (this.chars[charIndex].cst == 20) {
        hM = 5;
      }
    }
    return hitPoints + hM <= 0 && this.attacked[charIndex];
  };

  equipArmor = (armorType, weight, character) => {
    switch (armorType) {
      case ARMOR_TYPE.Bronze:
        character.arm = character.arm - 1;
        break;
      case ARMOR_TYPE.Steel:
        character.arm = character.arm + 1;
        break;
    }

    if (weight > 50) {
      character.arm = character.arm + 2;
    }
  };

  applyStrength = (strength, character) => {
    var charIndex = this.chars.indexOf(character);
    this.chars[charIndex].str = strength;
  };

  applyDexterity = (dexterity, character) => {
    var charIndex = this.chars.indexOf(character);
    this.chars[charIndex].dex = dexterity;
  };

  applyConstitution = (constitution, character) => {
    var charIndex = this.chars.indexOf(character);
    this.chars[charIndex].cst = constitution;
  };

  load = (filePath) => {
    if (filePath.endsWith('.json')) {
      return new Promise((resolve, reject) => {
        fs.readFile(filePath, { encoding: 'utf8' }, (err, contents) => {
          const json = JSON.parse(contents);
          const characters = json.characters;
          for (let i = 0; i < characters.length; i++) {
            this.chars[i] = new Character(characters[i].name, 5, characters[i].arm, characters[i].str, characters[i].dex, characters[i].const);
          }
          resolve();
        });
      });
    } else if (filePath.endsWith('.db')) {
      return new Promise(async (resolve, reject) => {
        const db = await sqlite.open(filePath, { Promise });
        const rows = await db.all('select Name, Armor, Str, Dex, Const from [Characters]');
        for (let i = 0; i < rows.length; i++) {
          this.chars[i] = new Character(rows[i].Name, 5, rows[i].Armor, rows[i].Str, rows[i].Dex, rows[i].Const);
        }
        await db.close();
        resolve();
      });
    } else {
      return new Promise((resolve, reject) => {
        fs.readFile(filePath, { encoding: 'utf8' }, (err, contents) => {
          const lines = contents.split('\n');
          for (let i = 1; i < lines.length; i++) {
            const values = lines[i].split(',');
            this.chars[i - 1] = new Character(values[0], 5, parseInt(values[1]), parseInt(values[2]), parseInt(values[3]), parseInt(values[4]));
          }
          resolve();
        });
      });
    }
  };
}
